### ISSUE TITLE
Seeing elevated network read ECONNRESET from AWS

### ISSUE THREAD

**Issue by nodesocket** [npm/registry-issue-archive#10]
Been seeing lots of npm `ERR! network read ECONNRESET` when invoking `npm install`.

Does this indicate a CDN issue on npm's side? This is from multiple AWS instances in the `us-west-1` and `us-west-2` regions.

```
npm ERR! network read ECONNRESET
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
npm ERR! Linux 3.13.0-48-generic
npm ERR! argv "node" "/home/node/.nvm/versions/node/v0.12.7/bin/npm" "install" "--production"
npm ERR! node v0.12.7
npm ERR! npm  v2.14.3
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! syscall read
```

**Comment by nodesocket** [comment]
In tandem also seeing:

```
npm ERR! Linux 3.13.0-48-generic
npm ERR! argv "node" "/home/node/.nvm/versions/node/v0.12.7/bin/npm" "install" "--production"
npm ERR! node v0.12.7
npm ERR! npm  v2.14.3

npm ERR! Callback called more than once.
npm ERR!
npm ERR! If you need help, you may report this error at:
npm ERR!     <https://github.com/npm/npm/issues>

npm ERR! Please include the following file with any support request:
npm ERR!     npm-debug.log
```

**Comment by chrisdickinson** [comment]
I'm looking into this — our CDN isn't reporting any issues. Would you mind sharing the output of `traceroute registry.npmjs.org`?

**Comment by blopker** [comment]
See also: https://github.com/npm/npm/issues/12484

**Comment by othiym23** [comment]
@blopker: I think you mean npm/npm#12484 – this is a separate repository.

**Comment by blopker** [comment]
Yep, already fixed it 😄

**Comment by nodesocket** [comment]
@chrisdickinson sorry for the delay, here is the result of traceroute:

```
$ traceroute registry.npmjs.org
traceroute to registry.npmjs.org (23.235.47.162), 30 hops max, 60 byte packets
 1  ec2-50-112-0-164.us-west-2.compute.amazonaws.com (50.112.0.164)  1.648 ms  1.662 ms  1.686 ms
 2  100.64.1.53 (100.64.1.53)  1.031 ms 100.64.1.63 (100.64.1.63)  1.024 ms 100.64.1.9 (100.64.1.9)  1.332 ms
 3  100.64.0.70 (100.64.0.70)  1.339 ms 100.64.0.96 (100.64.0.96)  1.357 ms 100.64.0.34 (100.64.0.34)  1.159 ms
 4  100.64.16.111 (100.64.16.111)  0.401 ms 100.64.16.159 (100.64.16.159)  0.418 ms 100.64.16.97 (100.64.16.97)  0.415 ms
 5  54.239.48.190 (54.239.48.190)  0.711 ms 54.239.48.188 (54.239.48.188)  0.734 ms 54.239.48.190 (54.239.48.190)  0.822 ms
 6  205.251.232.144 (205.251.232.144)  0.854 ms 205.251.232.162 (205.251.232.162)  0.718 ms 205.251.232.144 (205.251.232.144)  0.672 ms
 7  54.239.42.7 (54.239.42.7)  21.606 ms 205.251.232.109 (205.251.232.109)  19.991 ms 54.239.42.7 (54.239.42.7)  21.312 ms
 8  205.251.229.174 (205.251.229.174)  19.945 ms  20.479 ms 205.251.229.138 (205.251.229.138)  19.295 ms
 9  eqix-sv1.fastly.sv5-2.com (206.223.116.70)  19.626 ms 54.240.242.69 (54.240.242.69)  19.374 ms 54.240.242.73 (54.240.242.73)  19.969 ms
10  eqix-sv1.fastly.sv5-2.com (206.223.116.70)  19.740 ms * *
11  * * *
12  * * *
13  * * *
14  * * *
15  * * *
16  * * *
17  * * *
18  * * *
19  * * *
20  * * *
21  * * *
22  * * *
23  * * *
24  * * *
25  * * *
26  * * *
27  * * *
28  * * *
29  * * *
30  * * *
```

Also did mtr:

```
$ mtr --report registry.npmjs.org
Start: Fri Apr 29 17:07:40 2016
HOST: ip-172-31-14-205            Loss%   Snt   Last   Avg  Best  Wrst StDev
  1.|-- ???                       100.0    10    0.0   0.0   0.0   0.0   0.0
  2.|-- ???                       100.0    10    0.0   0.0   0.0   0.0   0.0
  3.|-- ???                       100.0    10    0.0   0.0   0.0   0.0   0.0
  4.|-- 100.64.16.23               0.0%    10    0.4   0.4   0.4   0.5   0.0
  5.|-- 54.239.48.188              0.0%    10    0.8   1.1   0.7   1.8   0.0
  6.|-- 205.251.232.144            0.0%    10    1.0   0.7   0.5   1.3   0.0
  7.|-- 205.251.232.109            0.0%    10   27.3  21.5  19.8  28.5   3.3
  8.|-- 205.251.229.170            0.0%    10   19.5  20.2  19.5  21.2   0.0
  9.|-- eqix-sv1.fastly.sv5-1.com  0.0%    10   19.8  19.9  19.8  19.9   0.0
 10.|-- ???                       100.0    10    0.0   0.0   0.0   0.0   0.0
 11.|-- 23.235.47.162              0.0%    10   19.9  20.1  19.9  20.7   0.0
```

**Comment by achaudhry** [comment]
Guys, I've pretty much lost all my hair trying to figure this out. Finally stumbled upon this issue. Any progress? I can't deploy anything on any of my EC2 instances. Thank you in advance! 

```
npm ERR! Linux 4.4.5-15.26.amzn1.x86_64
  npm ERR! argv "/opt/elasticbeanstalk/node-install/node-v4.3.0-linux-x64/bin/node" "/opt/elasticbeanstalk/node-install/node-v4.3.0-linux-x64/bin/npm" "--production" "install"
  npm ERR! node v4.3.0
  npm ERR! npm  v2.14.12
  npm ERR! code ECONNRESET
  npm ERR! errno ECONNRESET
  npm ERR! syscall read

  npm ERR! network read ECONNRESET
  npm ERR! network This is most likely not a problem with npm itself
  npm ERR! network and is related to network connectivity.
  npm ERR! network In most cases you are behind a proxy or have bad network settings.
  npm ERR! network 
  npm ERR! network If you are behind a proxy, please make sure that the
  npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
```

**Comment by nodesocket** [comment]
@achaudhry what AWS region are you in?

I've been trying to debug this all day with no luck. We are in `us-west-1` and `us-west-2`. I've tried increasing the SSD EBS volume size to 300GB to improve the IOPS, with no luck. Even have EBS optimized instances with type `c4.4xlarge` which should have plenty of cpu processing and memory.

It seems to be intermittent, though more often than not.

**Comment by achaudhry** [comment]
@nodesocket My instances are in us-west-2.

**Comment by achaudhry** [comment]
Although I wasn't actually going to use `http`, I tried changing the registry to `http` just out of curiosity but that didn't work for me either. Just FYI...

**Comment by nodesocket** [comment]
@achaudhry is it 100% error or does it work sometimes? Also, what instance type are you using? Using SSD EBS?

**Comment by achaudhry** [comment]
@nodesocket Worked once about 5 hours ago but 100% error since the last 4ish hours which is bizarre! I've tried on a couple of instances about 5-7 times on each. 

I have nothing fancy at the moment. Both instances are `t1.micro` - don't laugh, I'm a poor entrepreneur!

**Comment by nodesocket** [comment]
Your instance size may be the problem, 't1.micro' barely have any power and I/O. Would you be able to try a 'c4.large' just for a few minutes? Should only cost  15 cents or so.

I've tried 't2.large', 'c4.2xlarge' and even 'c4.4xlarge'.

**Comment by tolgaek** [comment]
Having the same problem on EC2 instance through Elastic Beanstalk

**Comment by soldair** [comment]
We are discussing internally how best to get enough data to figure out the issue. 

We are successfully serving so many requests from all zones of ec2 we know its not a general registry issue. This makes it very tricky to move forward but we are kicking around some ideas. Hope to have some more information soon. 

If anyone comes up with a way to profile their requests to the registry and finds out more please let us know.

**Comment by tolgaek** [comment]
I created a forum post on aws: https://forums.aws.amazon.com/thread.jspa?threadID=230574&tstart=0

Maybe that can get some attention from AWS side

**Comment by acusti** [comment]
Quoting a suggestion from @othiym23 in https://github.com/npm/npm/issues/9418#issuecomment-215605832 that hasn’t been mentioned yet in this thread:

> This is a network issue that occurs somewhere between the CLI and the CDN, and may or may not be due to the large number of simultaneous HTTPS connections initiated by the CLI. If it's the latter, than updating to the newest versions of npm@2 or npm@3 might mitigate the problem somewhat, due to the fact that they now limit the number of sockets open simultaneously.

**Comment by tolgaek** [comment]
Not sure if anything changed on AWS or registry side but I was just able to deploy 3 times in a row

**Comment by nodesocket** [comment]
Just confirmed that `npm install` is magically working again. This leads me to believe one of the two:
- 1.) The problem is npm or npm's CDN (Fastly) related.
- 2.) The problem is AWS related.

I'm leaning toward **1**, because AWS instances were up and running and seem to be working correctly all besides npm.

**Comment by achaudhry** [comment]
@nodesocket Agreed, it worked for me as well on the micro instance. Confused as to what was going wrong for two days straight. Would love to find out for future reference since that was not a very fun exercise...

That said, I'm happy to test it on a better / slightly more powerful instance but since it's working now, I highly doubt the instance type was the issue? Thanks!

**Comment by tolgaek** [comment]
@achaudhry I was blocked on very powerful instances. I don't think the issue was related to that

**Comment by sicloudhosting** [comment]
traceroute from c3.2xlarge in us-west-2

traceroute to registry.npmjs.org (199.27.79.162), 30 hops max, 60 byte packets
 1  ec2-50-112-0-164.us-west-2.compute.amazonaws.com (50.112.0.164)  1.067 ms ec2-50-112-0-198.us-west-2.compute.amazonaws.com (50.112.0.198)  0.963 ms ec2-50-112-0-164.us-west-2.compute.amazonaws.com (50.112.0.164)  1.027 ms
 2  100.64.1.207 (100.64.1.207)  1.023 ms 100.64.1.11 (100.64.1.11)  1.992 ms 100.64.1.223 (100.64.1.223)  1.188 ms
 3  100.64.0.38 (100.64.0.38)  1.102 ms 100.64.0.130 (100.64.0.130)  1.292 ms 100.64.0.228 (100.64.0.228)  1.922 ms
 4  100.64.16.211 (100.64.16.211)  0.305 ms 100.64.16.11 (100.64.16.11)  0.334 ms 100.64.16.77 (100.64.16.77)  0.314 ms
 5  54.239.48.188 (54.239.48.188)  0.613 ms 205.251.230.124 (205.251.230.124)  0.621 ms 54.239.48.188 (54.239.48.188)  0.568 ms
 6  205.251.232.162 (205.251.232.162)  0.694 ms 205.251.232.144 (205.251.232.144)  0.758 ms 205.251.232.150 (205.251.232.150)  0.888 ms
 7  205.251.232.93 (205.251.232.93)  8.024 ms 205.251.232.95 (205.251.232.95)  8.534 ms  7.906 ms
 8  52.95.52.24 (52.95.52.24)  7.245 ms 52.95.52.56 (52.95.52.56)  17.797 ms 52.95.52.192 (52.95.52.192)  12.497 ms
 9  52.95.52.53 (52.95.52.53)  7.815 ms 52.95.52.119 (52.95.52.119)  7.841 ms 52.95.52.157 (52.95.52.157)  8.028 ms
10  be-128-pe04.seattle.wa.ibone.comcast.net (66.208.228.209)  8.370 ms  7.358 ms be-127-pe04.seattle.wa.ibone.comcast.net (50.248.119.141)  8.264 ms
11  hu-1-2-0-9-cr01.seattle.wa.ibone.comcast.net (68.86.84.45)  8.972 ms hu-1-2-0-5-cr01.seattle.wa.ibone.comcast.net (68.86.82.229)  8.953 ms hu-0-7-0-4-cr02.seattle.wa.ibone.comcast.net (68.86.84.53)  9.475 ms
12  be-12125-cr01.9greatoaks.ca.ibone.comcast.net (68.86.85.198)  26.031 ms  25.998 ms be-10821-cr01.seattle.wa.ibone.comcast.net (68.86.85.81)  8.472 ms
13  be-10925-cr01.sunnyvale.ca.ibone.comcast.net (68.86.87.157)  27.973 ms  27.782 ms be-12125-cr01.9greatoaks.ca.ibone.comcast.net (68.86.85.198)  26.605 ms
14  be-10915-cr02.losangeles.ca.ibone.comcast.net (68.86.86.98)  31.558 ms  30.357 ms  32.436 ms
15  hu-0-2-0-1-pe02.600wseventh.ca.ibone.comcast.net (68.86.88.14)  30.807 ms be-10915-cr02.losangeles.ca.ibone.comcast.net (68.86.86.98)  32.551 ms  32.511 ms
16  hu-0-2-0-1-pe02.600wseventh.ca.ibone.comcast.net (68.86.88.14)  30.846 ms  30.992 ms  31.025 ms
17  \* \* *
18  \* \* *
19  \* \* *
20  \* \* *
21  \* \* *
22  \* \* *
23  \* \* *
24  \* \* *
25  \* \* *
26  \* \* *
27  \* \* *
28  \* \* *
29  \* \* *
30  \* \* *

**Comment by blopker** [comment]
Don't know if it helps, but we're getting this error on Centos 6, but not 5 or 7.

**Comment by nodesocket** [comment]
@chrisdickinson just started happening again. Seeing elevated connection timeouts consistently.

Region: `us-west-2`
Instance Type: `t2.large`

This is seriously affecting us, since all our of CI builds are failing.

```
npm ERR! network read ECONNRESET
npm ERR! network This is most likely not a problem with npm itself
npm ERR! network and is related to network connectivity.
npm ERR! network In most cases you are behind a proxy or have bad network settings.
npm ERR! network
npm ERR! network If you are behind a proxy, please make sure that the
npm ERR! network 'proxy' config is set properly.  See: 'npm help config'
npm ERR! Linux 3.13.0-48-generic
npm ERR! argv "node" "/home/node/.nvm/versions/node/v0.12.7/bin/npm" "install" "--production"
npm ERR! node v0.12.7
npm ERR! npm  v2.14.3
npm ERR! code ECONNRESET
npm ERR! errno ECONNRESET
npm ERR! syscall read
```

**Comment by tolgaek** [comment]
Can confirm I'm once again having issues as well :(

**Comment by thesmart** [comment]
This issue is intermittent in our us-west-2 ec2 clusters, occurring daily and maybe for a few minutes at a time. Kicking the build a second time usually does the trick. Considering the number of people reporting issue, I doubt it's a local problem unless maybe something is effed with amazons network. There seem to be a high number of Amazon users reporting in.

**Comment by terinjokes** [comment]
I wonder if this is a rate limit kicking in from Fastly, the reports in the last day all seem to be from AWS us-west.

**Comment by ricoli** [comment]
having the same issue on eu-west-1

**Comment by tolgaek** [comment]
started happening again. Anyone know of any workarounds until this is resolved? Can't get builds out!

**Comment by soldair** [comment]
@tolgaek are you running the most recent version of npm 2 or 3?
we are not overloaded. the response code count and request rates have not changed in any meaningful way.

what I'm thinking is if we run a little server on your boxes that proxies to the registry i can log each failure along with request duration and concurrency to the registry and maybe we can find a pattern. I'm going to try and make something quick. ill link it here so y'all can add things or give ideas.

**Comment by tolgaek** [comment]
@soldair I can install using your proxy. I'm using npm version `3.5.1`. Thanks for the help

**Comment by nodesocket** [comment]
@soldair we are running `2.14.3`.

Seeing from `us-west-1` and `us-west-2`.

**Comment by soldair** [comment]
@tolgaek & anyone willing to try the stats collecting proxy

`sudo npm install stats-npm -g`
https://www.npmjs.com/package/stats-npm

you can use either the server or replace calls to `npm` with `snpm` which is the best.

I have been using it for all of my npm installs today and i feel like most of the bugs are shaken out but please let me know if you have issues https://github.com/soldair/stats-npm/issues

its new software so i'm sure someone will hit a glaringly obvious bug an i apologize in advance. I think this is the only way to start figuring this out though.

**Comment by nodesocket** [comment]
Here is a traceroute from a instance that just failed. Don't see anything crazy.

```
$ traceroute registry.npmjs.org
traceroute to registry.npmjs.org (199.27.79.162), 30 hops max, 60 byte packets
 1  ec2-50-112-0-100.us-west-2.compute.amazonaws.com (50.112.0.100)  1.705 ms  1.704 ms  1.718 ms
 2  100.64.7.181 (100.64.7.181)  1.312 ms 100.64.7.191 (100.64.7.191)  0.945 ms 100.64.7.139 (100.64.7.139)  0.936 ms
 3  100.64.0.46 (100.64.0.46)  1.111 ms 100.64.0.198 (100.64.0.198)  1.036 ms 100.64.0.102 (100.64.0.102)  1.278 ms
 4  100.64.16.147 (100.64.16.147)  3.770 ms 100.64.16.211 (100.64.16.211)  3.766 ms 100.64.16.227 (100.64.16.227)  0.415 ms
 5  54.239.48.178 (54.239.48.178)  1.262 ms 54.239.48.176 (54.239.48.176)  1.134 ms 205.251.232.166 (205.251.232.166)  1.123 ms
 6  205.251.232.148 (205.251.232.148)  1.240 ms 205.251.232.154 (205.251.232.154)  1.188 ms 205.251.232.148 (205.251.232.148)  1.052 ms
 7  54.239.41.249 (54.239.41.249)  7.602 ms 54.239.41.253 (54.239.41.253)  8.184 ms  8.158 ms
 8  52.95.52.232 (52.95.52.232)  8.975 ms 52.95.52.196 (52.95.52.196)  9.231 ms 52.95.52.90 (52.95.52.90)  13.838 ms
 9  52.95.52.89 (52.95.52.89)  8.417 ms 52.95.52.49 (52.95.52.49)  8.596 ms 52.95.52.51 (52.95.52.51)  8.548 ms
10  be-106-pe05.seattle.wa.ibone.comcast.net (50.242.148.93)  11.282 ms be-107-pe05.seattle.wa.ibone.comcast.net (50.242.148.89)  11.288 ms be-106-pe05.seattle.wa.ibone.comcast.net (50.242.148.93)  6.971 ms
11  hu-0-0-0-5-cr02.seattle.wa.ibone.comcast.net (68.86.89.117)  9.961 ms hu-1-2-0-0-cr01.seattle.wa.ibone.comcast.net (68.86.85.57)  8.991 ms hu-0-0-0-3-cr02.seattle.wa.ibone.comcast.net (68.86.89.85)  10.091 ms
12  be-12125-cr01.9greatoaks.ca.ibone.comcast.net (68.86.85.198)  24.886 ms be-10821-cr01.seattle.wa.ibone.comcast.net (68.86.85.81)  9.917 ms be-12125-cr01.9greatoaks.ca.ibone.comcast.net (68.86.85.198)  24.888 ms
13  be-12125-cr01.9greatoaks.ca.ibone.comcast.net (68.86.85.198)  24.906 ms  24.888 ms be-10925-cr01.sunnyvale.ca.ibone.comcast.net (68.86.87.157)  22.842 ms
14  be-10925-cr01.sunnyvale.ca.ibone.comcast.net (68.86.87.157)  23.783 ms be-10915-cr02.losangeles.ca.ibone.comcast.net (68.86.86.98)  33.471 ms  32.517 ms
15  hu-0-0-0-0-pe02.600wseventh.ca.ibone.comcast.net (68.86.83.18)  31.399 ms be-10915-cr02.losangeles.ca.ibone.comcast.net (68.86.86.98)  33.218 ms hu-0-0-0-0-pe02.600wseventh.ca.ibone.comcast.net (68.86.83.18)  31.314 ms
16  * hu-0-0-0-0-pe02.600wseventh.ca.ibone.comcast.net (68.86.83.18)  31.569 ms *
17  * * *
18  * * *
19  * * *
20  * * *
21  * * *
22  * * *
23  * * *
24  * * *
25  * * *
26  * * *
27  * * *
28  * * *
29  * * *
30  * * *
```

**Comment by soldair** [comment]
yeah i dont think we're getting anything helpful from traceroute sadly. let me know if you try stats-npm and have a snpm-stats.log file though.

**Comment by nodesocket** [comment]
I am going to try `stats-npm`, but looking at the `npm-debug.log` I do see:

```
134616 verbose stack Error: read ECONNRESET
134616 verbose stack     at exports._errnoException (util.js:746:11)
134616 verbose stack     at TCP.onread (net.js:559:26)
```

Are those lines helpful at all?

**Comment by soldair** [comment]
happy to look at it i dont think so though.
its not enough to just know its happening i need a bit more context like how many other requests were active for which kinds of resources and how long was the request active before the reset.

if you make any snpm-stats.logs send the debug log with it to support@npmjs.com and ill take a look.

**Comment by nodesocket** [comment]
@soldair here it is. Saws lot of `read ECONNRESET` errors. Let me know what else I can do to help.

Gist: https://gist.github.com/nodesocket/9d7188f09d37614911aad8e31e346fc7

**Comment by soldair** [comment]
Thanks!

**Comment by nodesocket** [comment]
I just tried bumping to the latest `0.12.x` so `0.12.13` which bundles npm `2.15.0`.

With this new version I was able to successfully run `npm install` 10 times in a row without seeing any `ECONNRESET` errors on an AWS instance.

The tricky part is I don't know if the updated node/npm version fixed it, or just I got lucky.

**Comment by deleteme** [comment]
Is https://github.com/nodejs/node/issues/3595 related?

**Comment by achaudhry** [comment]
@nodesocket I hope so man although I have node 4.3.0 and npm 2.14.12. I guess I can try npm 2.15.0 and see if that makes a difference...

**Comment by soldair** [comment]
### please do this stuff.
1. upgrade npm!
2. i need more snpm-stats.logs
3. all of the requests npm sent worked.
4. it still got ECONNRESET
5. it may be keep-alive connections in the pool
6. the reason its throwing these is probably https://github.com/nodejs/node/issues/3595 (recently fixed in 4.x not sure if it's released)

### TL;DR;

here is what i'd like folks to do.
- Please upgrade your npm. as mentioned already in this thread. The cli team recommends this. With improved pooling logic these newest versions are more reliable.

`npm -g install npm` for 3 or `npm -g install npm@latest-2` for 2
- use stats-npm for collecting networking stats when you install. this write an append log so it will keep collecting with each install.
  `npm  -g install snpm` and replace calls directly to `npm`  with `snpm`
-  upload both snpm-stats.log and npm-debug.log as gists and please send them to support@npmjs.com.

from @nodesocket 's snpm-stats.log  It seems like ECONNRESET are not happening to the actual requests. All of the proxied requests completed successfully on the first try but they claimed that the install failed due to the error anyway. 

I believe @deleteme is on the right path. this very well may be https://github.com/nodejs/node/issues/3595 and or generally keep-alive related. the stats-npm proxy replaces `connection: keep-alive` with `connection: close` on outgoing requests because i had issues with keep alive pooling locking up the global agent. The only place where keep-alive type pooling is going on is _in the npm node process itself_.

ECONNRESET happens in a number of cases. sometimes when the opposite side of the connection forceably closes the tcp connection but also in cases where something happens in IP routing land and the fin packet never makes it to the client.

**Comment by tolgaek** [comment]
I think this might be `Elastic Beanstalk` related. First off @soldair with EB deploy, failed deploy folders are removed so I can't find those log files. Anyone know where I can find the `snpm` logs?

More importantly, I have done an npm install on the same EC2 instance using ssh and was able to npm install about 5 times in a row without any issues.

But as soon I tried a deploy through beanstalk, it failed. Although looking at logs, I was running npm install with exact same npm and node versions, it was still somehow working when directly run. Not sure how those 2 are different but I think it might be related

**Comment by soldair** [comment]
snpm drops the log in the current working directory of the process. this should be the top level of the project as this is where you normally run npm install. I'll be updating stats-npm to upload the stats as soon as i get a chance so we can still get the information we need to debug. 

@tolgaek if you review my comment ^ from the snpm-stats.logs I have already npm cli doesnt fail any requests is actually makes. now there also make be legit networking issues in beanstalk somehow which means getting these logs is the most important thing right now. ill ping this thread when i have it. 

snpm also will retry econnreset errors on sockets 4 times before giving up and failing to npm cli with a 500. so using the snpm command it would be incredibly unlikely to get a real econnreset from 127.0.0.1

**Comment by ricoli** [comment]
I stopped having the ECONNRESET problems after I started using npm 3.8.7 (and I was having ECONNRESET all the time on CI server, so I don't think it's anything purely related to Elastic beanstalk @tolgaek )

**Comment by tolgaek** [comment]
I see thanks @soldair I will still try to dig up the logs but beanstalk seems to delete the whole folder when a deploy fails.

@ricoli yeah I'm on 3.8.8, my deploys aren't failing right now but I will try to deploy with 3.8.7 when it happens again

**Comment by ricoli** [comment]
For reference, I was having all the ECONNRESET problems when using 3.5.2 (quite old I know... it was time to upgrade :D )

**Comment by soldair** [comment]
is anyone experiencing this problem using a node version 
`>=  4.4.0 , and any 5 or 6`

a fix related to out of band errors from keep-alive pooling should have been included in all of these versions. trying to figure out if this node bug is the culprit.

**Comment by ricoli** [comment]
@soldair I was having this issue using node 5.3 and npm 3.5.2. No longer having it when using node 6.0 and npm 3.8.7

**Comment by soldair** [comment]
@ricoli digging around the fix may have been shipped to 5 in 5.4

**Comment by nodesocket** [comment]
@soldair I'm not entirely convinced that upgrading to the latest version will fix this. Why has this worked solid for over 6 months, but then all the sudden started breaking. Does that make sense?

**Comment by soldair** [comment]
im not convinced of anything =). 
all we can do is chase down leads. folks have been reporting econnreset for years AFAIK this is not a new thing.

i believe this at it's core is a bug in node that manifests itself inconsistently based on hand wavy host conditions. with your very own test data we confirmed that this is NOT an external network issue. the econnresets happen without keep alive and every outgoing request made by the npm process that really leaves the process works.

the change in newer versions of node handles this error when it happens out of band to a request. this probably will not stop undiscovered bad sockets from being served from the pool to http clients using keep-alive pooling.

**Comment by codeimpossible** [comment]
**_tl;dr_ Upgrading to node 4.4.x+ solved the problem for us. We haven't tried node versions > 5.x yet.**

All of our build agents at Hudl were experiencing ECONNRESET errors when communicating with the registry over SSL. The failures weren't consistent, and would crop up at seemingly random times. 

We narrowed the failure down to versions of node less than 4.2.x after a few weeks of fixing/testing pretty much anything else. After some digging we found this issue in the node repo: https://github.com/nodejs/node/issues/3595.

We confirmed that this was our problem by upgrading our version of node to 4.4.2 on our build agents and developer machines. Once we did this the error stopped happening completely. We verified the fix by downgrading and seeing the error happen, then upgrading and seeing the error disappear.

**Comment by khisakuni** [comment]
I'm having this same issue when trying to deploy AWS Elastic Beanstalk. Are there any suggestions on getting around this in a situation where I can't upgrade the node version? The latest node version that can be used on elastic beanstalk is 4.3.0

**Comment by soldair** [comment]
@khisakuni im not experienced with beanstalk but with this approach you should be able to use a simple script to download and install a newer version of nodejs and npm.

http://stackoverflow.com/questions/31260116/how-do-i-run-npm-update-g-npm-on-elastic-beanstalk

install script
https://gist.github.com/TooTallNate/3288316

**Comment by soldair** [comment]
im closing this issue as it seems like the fix that made it into node >=4.4 >=5.4 and >= 6 solves the issue.

**Comment by deleteme** [comment]
I can confirm that upgrading node from `5.1.1` to `5.11.0` fixes it.

**Comment by blopker** [comment]
Just one more datapoint: 4.4.3 fixed it for us too.

**Comment by adrianovalente** [comment]
I had the same problem deploying to aws eb with npm 2.4.12. Changing registry to `http` worked for me!

**Comment by achaudhry** [comment]
@nodesocket @soldair so after about 5 months all of a sudden the issue has resurfaced for me and it's happening consistently today. Anyone seeing it again? I remember last time I tried everything mentioned above and nothing worked for me but it randomly started working one day. I've again tried everything mentioned here but so far no luck. Anyone figured out the root cause?

**Comment by soldair** [comment]
Node and npm client version?
To debug we need more data. If you install with stats-npm and send me the log for failures i stand the best chance to figure it out.

We haven't had and infrastructure or system changes recently that could cause this so we are in much the same place as before.

**Comment by atungare** [comment]
@soldair We are seeing a bunch of these ECONNRESETs recently on Travis, with e.g. node 5.8.0 and npm 3.7.3 (but other versions of node & npm are also experiencing errors).

**Comment by soldair** [comment]
@atungare we are not experiencing any downtime or service interuption as far as i can tell. we'll have to find a way to get stats-npm logs for your install at least an npm-debug.log to debug

**Comment by clareliguori** [comment]
My team is seeing this on AWS today (see https://github.com/npm/registry/issues/112).

I'll throw out a possibility: does npm registry (or Fastly) do any IP-based throttling?  It seems fairly consistent that builds run "in the cloud" (on AWS or on Travis) are affected, and I'm assuming their requests to npm registry would all come from the same IP range.  I'm wondering if rate limiting is kicking in.

**Comment by jboler** [comment]
About 50% of our builds on our paid Travis-CI account fail with this error. node 4.1.2, npm 2.15.11.

**Comment by pjar** [comment]
It's holding my work too. NPM is so broken - how the hell it became a kind of standard? 
Sorry, I just lost so much time on npm errors this week, getting frustrated...
